File: src/app/conjugation/forms/passive-affirmative.ts
import {Conjugation, Word, WordType} from "../conjugation";
import {HIRAGANA} from "../hiragana";

export class PassiveAffirmative implements Conjugation {
    getConjugation(word: Word): (Word | undefined) {
        switch (word.wordType) {

            // Verbs
            case WordType.GodanVerb:
                const lastKana = HIRAGANA[word.getLastKana()]
                if (lastKana.kana === 'う') {
                    return word.replaceLastKana('われる')
                }

                return word.replaceLastKana(lastKana.getGroup().a + 'れる')

            case WordType.IchidanVerb:
                return word.replaceLastKana('られる');

            case WordType.SuruVerb:
                return word.replaceLastKana('される', 2);

            case WordType.KuruVerb:
                return word.replace('来られる', 'こられる');


            default:
                return undefined;
        }
    }

    getTitle(): string {
        return "Passive, affirmative"
    }

    getSettingsTitle(): string {
        return "Affirmative"
    }

}

File: backend/app/conjugation/forms/passive_affirmative.py
from ..core import Conjugation, Word, WordType
from ..hiragana import HIRAGANA


class PassiveAffirmative(Conjugation):

    title = 'Passive, affirmative'
    settings_title = 'Affirmative'

    def conjugate(self, word: Word) -> Word | None:
        match word.word_type:

            # Verbs
            case WordType.GODAN_VERB:
                last_kana = HIRAGANA[word.last_kana]
                if last_kana.kana == 'う':
                    return word.replace_last_kana('われる')

                return word.replace_last_kana(last_kana.group.a + 'れる')

            case WordType.ICHIDAN_VERB:
                return word.replace_last_kana('られる')

            case WordType.SURU_VERB:
                return word.replace_last_kana('される', 2)

            case WordType.KURU_VERB:
                return word.replace('来られる', 'こられる')

            case _:
                return None
